### DISCUSSION TITLE
Does ib-api work with the latest version of IB Gateway?

### DISCUSSION THREAD

**Discussion by georgeu2000** [ib-ruby/ib-api#26] [General] [answered]
Currently IB Gateway is Version 10.19.1f .

Does ib-ruby work with it? If not, what is the latest version that works?

**Comment by topofocus** [comment, marked as answer]
Hi,
personally, I am using IB-gateway with TWS 985 and TWS 10.12.
New TWS-Versions mostly do not add much new functions neither change existing ones.
The main issue in V10x is the support of crypto-trading. 

Adaption of ib-api to V10x is on the way. On the "Ver 10" Branch there is an Issue on submitting orders left.

**Reply by georgeu2000** [reply]
Great. Thank you.

**Comment by georgeu2000** [comment]
@topofocus - Thanks for all you work on these gems.

I tried ver_10 branch [57ef882](https://github.com/ib-ruby/ib-api/commit/57ef8820edcfc9fdc6aa10238386526c33daa0f3) andIB Gateway 10.15.

Here is the error:

```
2[[20, 2, "", "YM", "CONTFUT", "", "", "", "", "ECBOT", "", "USD", "", "", 0, nil, "5 secs", "1000 S", 0, "TRADES", 2, 1, "XYZ"]]
 and request_id 2
TWS Error 366: No historical data query found for ticker id:1
E: Error reading request. Please use 'Key=Value' format for Misc Options id: 2
TWS Error 320: Error reading request. Please use 'Key=Value' format for Misc Options
```

I get the `RequestHistoricalData` for Futures contracts. Works fine with [c88da](https://github.com/ib-ruby/ib-api/commit/c88da8b9f8add57a539070378f2cd8fda26cb2e4) and IB Gateway 981.

```
request_id = @ib.send_message IB::Messages::Outgoing::RequestHistoricalData.new(
                      id: @last_ticker_id,
                      contract: contract,
                      duration: back_for(timeframe),
                      bar_size: timeframe,
                      what_to_show: :trades,
                      use_rth: 0, # 0 => include hours outside of normal hours
                      keep_up_todate: keep_up_todate) # 1 => true, 0 => false
```

Do I need to change something?

Thank you!

**Comment by topofocus** [comment]
Will look to this.

Can you please try the eod method first?

```ruby

require 'ib-api' 
require 'ib-extensions'  #  or ib-gateway 
IB::Stock.new( symbol: 'GE' ).
           eod( start: Date.new(2015,6,15), to: Date.new(2015,6,21) ).
           map( &:close ).
           join( " : " )

```
This works as expected in 10.12. 

expected Result: "209.49 : 209.33 : 209.95 : 209.41"

btw:  do not specify 'id' if you expect a request-id as return value.

**Comment by georgeu2000** [comment]
I got 
```
undefined method `eod' for #<IB::Stock:0x000000010b39b7a8 @attributes={:symbol=>"GE", :created_at=>2022-10-31 15:18:03.557005 -0700, 
:con_id=>0, :right=>"", :include_expired=>false, :sec_type=>"STK", :currency=>"USD", :exchange=>"SMART"}> (NoMethodError)`
```

**Comment by topofocus** [comment]
Please include 
```
gem "ib-api"  , git: "https://github.com/ib-ruby/ib-api"
gem "ib-symbol, git: "https://github.com/ib-ruby/ib-symbol"
gem "ib-extentions", git: "https://github.com/ib-ruby/ib-extensions"
```
in your gemfile

and 

'require "ib-api"`
`require "ib-extensions"`
to your script.

alternatively you can clone ib-extensions  and use the inlcuded bin/console  / bin/gateway  to try the eod-method.

It returns the results and stores them for convenience in contract.bars. 

its documentated: https://ib-ruby.github.io/ib-doc/Historical_data.html

According to my test, using the stable-tws/gateway release 10.12, it works perfectly, either  with ib-api- gem and  the  current 'v10' branch of ib-api.

**Comment by georgeu2000** [comment]
Hi,

Sorry for the late response.

I created a new project and set up the test script again.

```
$ cat Gemfile
source "https://rubygems.org"
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby '3.1.2'

gem "ib-api"  , git: "https://github.com/ib-ruby/ib-api"
gem "ib-symbols", git: "https://github.com/ib-ruby/ib-symbols"
gem "ib-extensions", git: "https://github.com/ib-ruby/ib-extensions"
```

```
$ cat ib-test.rb
require 'ib-api'
# require 'ib-extensions'  #  or ib-gateway
require 'ib-gateway'

IB::Stock.new( symbol: 'GE' ).
           eod( start: Date.new(2015,6,15), to: Date.new(2015,6,21) ).
           map( &:close ).
           join( " : " )
```

```
$ bundle exec gem list ib-

*** LOCAL GEMS ***

ib-api (972.5.2)
ib-extensions (1.3.1)
ib-symbols (1.4)
```


```
 bundle exec ruby ib-test.rb
/Users/george/.rbenv/versions/3.1.2/gemsets/ib-api-test/bundler/gems/ib-extensions-b3e0483edce9/lib/ib/eod.rb:158:in `block in eod': undefined method `[]' for nil:NilClass (NoMethodError)

            if d.is_a?(Integer) || !["D","M","W","Y"].include?( d[-1].upcase )
                                                                 ^^^^
	from /Users/george/.rbenv/versions/3.1.2/gemsets/ib-api-test/bundler/gems/ib-extensions-b3e0483edce9/lib/ib/eod.rb:195:in `eod'
	from ib-test.rb:6:in `<main>'
```

I was not able to require 'ib-extensions'. It raised an error.

So, is there a better way to test the ib-api functionality?

Which version of Ruby are you using? I am using 3.1.2.

Which version of IB Gateway are you using?

Thank you once again for the awesome gems!!!

**Reply by topofocus** [reply]
Hi,
ruby 3.1.2 is fine. You can even use the jit-compiler

To fix your error, please try to load the latest version in the Gemfile

gem 'ib-api', :git=> 'https://github.com/ib-ruby/ib-api.git'
gem 'ib-extensions',  git: 'https://github.com/ib-ruby/ib-extensions.git'
gem 'ib-symbols'


then open a console
and your code should work as expected.

I have never seen this error. Thanks for reporting it.

**Comment by georgeu2000** [comment]
Hi,

Thanks for the quick response.

I think I am using the latest versions.

```
bundle
...

Using ib-api 972.5.2 from https://github.com/ib-ruby/ib-api (at master@00555fc)
Using ib-extensions 1.3.1 from https://github.com/ib-ruby/ib-extensions (at master@b3e0483)
Using ib-symbols 1.4 from https://github.com/ib-ruby/ib-symbols (at master@941b987)
```

Using the latest releases (by removing the `git: repo` from the Gemfile) has the same symptom.

Not sure what I am missing...

**Comment by georgeu2000** [comment]
While doing other testing, I found that the latest versions of the gems works with IB Gateway 10.15.

```
$ bundle | grep ib-
Using ib-api 972.5.2
Using ib-extensions 1.3.1
Using ib-symbols 1.5
```

So looks like things are sorted for now.

**Comment by topofocus** [comment]
thats puzzling.
Which TWS-Version did you use before?

just created a new branch `test_eod` in ib-extensions
can you please load this in your gem?

I added a check of a proper class of the duration.

**Comment by georgeu2000** [comment]
I was not using TWS, only IB Gateway 981.

OK, I ran with `test_eod`:

```
$ bundle | grep ib-
Using ib-api 972.5.2 from https://github.com/ib-ruby/ib-api
Using ib-extensions 1.3.1 from https://github.com/ib-ruby/ib-extensions (at test-eod@bb05028)
Using ib-symbols 1.4 from https://github.com/ib-ruby/ib-symbols
```

```
$ ruby ib-test.rb
/Users/george/.rbenv/versions/3.1.2/gemsets/ib-api-test/gems/ib-extensions-1.3.1/lib/ib/eod.rb:158:in `block in eod': undefined method `[]' for nil:NilClass (NoMethodError)

            if d.is_a?(Integer) || !["D","M","W","Y"].include?( d[-1].upcase )
```

This does not seem to be using the new branch. This error line appears to be 158 of the main or release branch.

```
$ cat ib-test.rb
require 'ib-api'
# require 'ib-extensions'  #  or ib-gateway
require 'ib-gateway'

IB::Stock.new( symbol: 'GE' ).
           eod( start: Date.new(2015,6,15), to: Date.new(2015,6,21) ).
           map( &:close ).
           join( " : " )
```

There is some issue with requiring 'ib-extensions'.
When I attempt to require it, I get

```
$ ruby ib-test.rb
<internal:/Users/george/.rbenv/versions/3.1.2/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:148:in `require': cannot load such file -- ib-extensions (LoadError)
```

But not sure what would cause that...maybe something in the Gemspec?

**Reply by topofocus** [reply]
Do not use the old TWS versions anymore. They are not supported by the IB-Server.  However, the access via the 9xx protocol is stable.
There were several api-changes. Maybe thats the issue. 
ib-ruby is now tested with TWS Version 10.12 to 10.20
